// instagrapi/dist/utils/instagram.util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstragramUtil = void 0;
var general_util_1 = require("./general.util");
var InstragramUtil = (function () {
    function InstragramUtil() {
    }
    InstragramUtil.getCaption = function (media) {
        var caption = media.edge_media_to_caption.edges.length
            ? media.edge_media_to_caption.edges[0].node.text
            : media.accessibility_caption || '';
        return caption ? caption : media.is_video ? media.title : null;
    };
    InstragramUtil.getComments = function (commentList, proxy) {
        if (commentList === null || commentList === void 0 ? void 0 : commentList.length) {
            return commentList.map(function (comment) {
                var user = comment.user;
                return {
                    content: comment.text,
                    author: {
                        username: user.username,
                        name: user.full_name,
                        image: proxy + user.profile_pic_url,
                        isVerified: user.is_verified,
                        isPrivate: user.is_private
                    },
                    likes: comment.comment_like_count,
                    isSpam: comment.did_report_as_spam,
                    date: general_util_1.GeneralUtil.msToDate(comment.created_at)
                };
            });
        }
        else
            return [];
    };
    InstragramUtil.getLocation = function (location) {
        if (location) {
            var _a = location.name.split(','), region = _a[0], country = _a[1];
            var locationWrapper = {
                country: (country === null || country === void 0 ? void 0 : country.trim()) || null,
                region: region || null,
                city: location.city || null,
                street: location.address || null,
                coordinates: {
                    lat: location.lat,
                    lng: location.lng
                }
            };
            return locationWrapper;
        }
        else
            return null;
    };
    InstragramUtil.getMediaQualities = function (media, proxy) {
        var _a;
        var imageWidth480 = (_a = media.find(function (img) { return img.width === 480; })) === null || _a === void 0 ? void 0 : _a.url;
        return {
            standard: proxy + (imageWidth480 || media.pop().url),
            hd: proxy + media.shift().url
        };
    };
    InstragramUtil.getAudioData = function (item) {
        var _a, _b;
        if (item === null || item === void 0 ? void 0 : item.has_audio) {
            var audioUser = (_a = item.clips_metadata) === null || _a === void 0 ? void 0 : _a.original_sound_info;
            var audioArtist = (_b = item.clips_metadata) === null || _b === void 0 ? void 0 : _b.music_info;
            if (audioUser) {
                return {
                    artist: audioUser.ig_artist.full_name,
                    song: null
                };
            }
            else if (audioArtist) {
                return {
                    artist: audioArtist.music_consumption_info.ig_artist.full_name,
                    song: audioArtist.music_asset_info.title
                };
            }
        }
        return null;
    };
    return InstragramUtil;
}());
exports.InstragramUtil = InstragramUtil;
